src/proxy.rs

```diff
@@ -1,74 +1,107 @@
+        let mut client_req = client::Request::new(req.method().clone(), url.parse().unwrap());
+
+        // TOOD put this in a separate function. it should filter out headers we should not pass
+        // we should also add headers like via
+        {
+            let mut client_headers = client_req.headers_mut();
+            client_headers.extend(req.headers().iter());
         }
-    }
-}
-
-pub fn listen(addr: SocketAddr, pool: Pool) -> result::Result<(), io::Error> {
 
-    let mut lp = Core::new().unwrap();
-    let handle = lp.handle();
+        // TODO store client in Proxy object
+        let client = Client::new(&self.handle.clone());
+        let backend = client.call(client_req).and_then(|res| {
+            debug!("Response: {}", res.status());
+            debug!("Headers: \n{}", res.headers());
 
-    let listener = TcpListener::bind(&addr, &handle.clone()).expect("Failed to bind to socket");
-    info!("Listening on: {}", addr);
+            // transform client::Response into server::Response
+            let mut server_response = Response::new().with_status(*res.status());
 
-    let f = listener.incoming().for_each(|(sock, addr)| {
-        debug!("Incoming connection on {}", addr);
+            {
+                let mut server_headers = server_response.headers_mut();
+                server_headers.extend(res.headers().iter());
+            }
 
-        let addr = pool.get().expect("Failed to get address from pool");
-        debug!("Preparing backend request to {:?}", addr);
+            server_response.set_body(res.body());
 
-        // We attach the handle to Frontend to force the compiler to acknowledge that `backend`
-        // will not outlive the handle.
-        let frontend = frontend::Frontend { handle: handle.clone() };
-        let client = TcpClient::new(backend::Backend);
-        let f = client.connect(&addr, &handle.clone()).map(move |backend| {
-            let service = Proxy::new(backend);
-            frontend.bind_server(&frontend.handle, sock, service);
+            ::futures::finished(server_response)
         }).map_err(|e| {
             error!("Error connecting to backend: {:?}", e);
-            ()
+            e
         });
 
-        handle.spawn(f);
+        Box::new(backend)
+    }
+}
 
-        Ok(())
-    });
+struct NewProxyService {
+    handle: Handle,
+    pool: Pool,
+}
+
+impl NewService for NewProxyService {
+    type Request = Request;
+    type Response = Response;
+    type Error = hyper::Error;
+    type Instance = Proxy;
+
+    fn new_service(&self) -> io::Result<Self::Instance> {
+        Ok(Proxy {
+            handle: self.handle.clone(),
+            pool: self.pool.clone(),
+        })
+    }
+}
+
+pub fn listen(addr: SocketAddr, pool: Pool) -> result::Result<(thread::JoinHandle<()>, Listening), io::Error> {
+    let (tx, rx) = ::std::sync::mpsc::channel();
 
-    lp.run(f).expect("Unexpected error while proxying connection");
+    let handle = thread::spawn(move || {
```

Comment:
Are we forced to use 2 threads with that (the calling thread and the thread for the server)?
Maybe we could have 2 `pub fn`, one returning a `JoinHandle` and one blocking the calling thread to avoid wasting one thread?